### PR TITLE
Add brew shellenv - see #4780

### PR DESCRIPTION
**Pull request by jamescostian**
- [x] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/brew/blob/master/CONTRIBUTING.md) document?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/brew/pulls) for the same change?
- [x] Have you added an explanation of what your changes do and why you'd like us to include them?
- [x] Have you written new tests for your changes? [Here's an example](https://github.com/Homebrew/brew/blob/master/Library/Homebrew/test/PATH_spec.rb).
- [x] Have you successfully run `brew style` with your changes locally?
- [x] Have you successfully run `brew tests` with your changes locally?

-----

Adds `brew shellenv` based on conversation in #4780

### CODE REVIEW

File: Library/Homebrew/test/cmd/shellenv_spec.rb [hunk 1]
```diff
@@ -0,0 +1,7 @@
+describe "brew shellenv", :integration_test do
+  it "doesn't fail" do
```
Comment (MikeMcQuaid) [inline Library/Homebrew/test/cmd/shellenv_spec.rb]: Could you `assert_match` on e.g. the `HOMEBREW_PREFIX/bin` being present?

**Comment by sjackman** [comment]
Looks good to me. Thanks, James! The one minor test change suggested by make would be good, to check that stdout contains `#{HOMEBREW_PREFIX}/bin`.

**Comment by sjackman** [comment]
After this is merge into Homebrew/brew and Linuxbrew/brew, we can update the installation instructions.

**Comment by jamescostian** [comment]
Just added that to the test 👍

**Comment by sjackman** [comment]
Merged. Thanks for your contribution to Homebrew, James! 🏆

**Comment by sjackman** [comment]
There's two ways that `brew shellenv` could be added to ~/.profile.
```sh
~/.linuxbrew/bin/brew shellenv >>~/.profile
```
or
```sh
echo 'eval $(~/.linuxbrew/bin/brew shellenv)' >>~/.profile
```
I have a slight preference for the latter using `eval`. Your opinion?

**Comment by jamescostian** [comment]
Running `time bin/brew shellenv` on my laptop (processor is i7-4870HQ, running at 2.8GHz, mostly idling) reports a real time of 0.40s, which IMO makes it way too slow for me to be willing to `eval` it more than once

**Comment by MikeMcQuaid** [comment]
@sjackman My recommendation would be to advise people put `eval` in their `profile` and those who care about the overhead can hardcode it if necessary.

Another possible extension for this script would be exporting e.g. `HOMEBREW_PREFIX` and `HOMEBREW_CELLAR` so scripts (or the profile itself) can avoid calling `brew --prefix` to further improve performance. 

Nice work @jamescostian!

**Comment by sjackman** [comment]
> My recommendation would be to advise people put eval in their profile and those who care about the overhead can hardcode it if necessary.

Is that so that we can change the output of `brew shellenv`, and have all users up-to-date without needing to modify their shell configuration?

I suspect that implementing this script in shell rather than Ruby would execute quite a bit faster. If it's going in `~/.profile` maybe it's not such a big deal, since it'd only be executed when opening a new terminal (or ssh session). I'd be more concerned if it's going in `~/.bashrc` where it'd be executed for every subshell.

**Comment by MikeMcQuaid** [comment]
> Is that so that we can change the output of brew shellenv, and have all users up-to-date without needing to modify their shell configuration?

Yep 👍 



> I suspect that implementing this script in shell rather than Ruby would execute quite a bit faster

It would indeed. Given it only relies on `HOMEBREW_PREFIX` being defined and even any extensions I see would only need things already defined in the Bash layer it should be possible to rewrite it in Bash, name it `cmd/shellenv.sh` and it'll avoid needing to ever execute the Ruby layer.

**Comment by sjackman** [comment]
@jamescostian Would you be interested in submitting a PR to rewrite `shellenv` in shell so that it takes less time to execute?

**Comment by jamescostian** [comment]
Sure, the bash rewrite is pretty simple:

```bash
echo "export PATH=\"$HOMEBREW_PREFIX/bin:$HOMEBREW_PREFIX/sbin:\$PATH\""
echo "export MANPATH=\"$HOMEBREW_PREFIX/share/man:\$MANPATH\""
echo "export INFOPATH=\"$HOMEBREW_PREFIX/share/info:\$INFOPATH\""
```

But there are two issues:

1. I'm getting an error at the end of the output: "Library/Homebrew/brew.sh: line 405: homebrew-shellenv: command not found" - any tips for that?
2. Even if I comment out the line that makes that error, the execution time is .10s

That execution time seems to be impacted most from [this line](https://github.com/Homebrew/brew/blob/master/Library/Homebrew/brew.sh#L262) and [this line](https://github.com/Homebrew/brew/blob/master/Library/Homebrew/brew.sh#L58). They both talk to `git` and they both look important enough to keep.

One way to decrease the execution time to .01s would be to treat `shellenv` completely differently:

```diff
--- a/bin/brew
+++ b/bin/brew
@@ -65,7 +65,7 @@ do
 done

 # test-bot does environment filtering itself
-if [[ -z "$HOMEBREW_NO_ENV_FILTERING" && "$1" != "test-bot" ]]
+if [[ -z "$HOMEBREW_NO_ENV_FILTERING" && "$1" != "test-bot" && "$1" != "shellenv" ]]
 then
   PATH="/usr/bin:/bin:/usr/sbin:/sbin"

@@ -82,6 +82,11 @@ then
   done

   exec /usr/bin/env -i "${FILTERED_ENV[@]}" /bin/bash "$HOMEBREW_LIBRARY/Homebrew/brew.sh" "$@"
+elif [[ "$1" == "shellenv" ]]
+then
+  echo "export PATH=\"$HOMEBREW_PREFIX/bin:$HOMEBREW_PREFIX/sbin:\$PATH\""
+  echo "export MANPATH=\"$HOMEBREW_PREFIX/share/man:\$MANPATH\""
+  echo "export INFOPATH=\"$HOMEBREW_PREFIX/share/info:\$INFOPATH\""
 else
   source "$HOMEBREW_LIBRARY/Homebrew/brew.sh"
 fi
```

But obviously that's not very elegant...

@MikeMcQuaid thoughts?

**Pull request merged**